_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+address to the Erlang stack followed by a `goto` to the label of the
+called function.  (Strictly speaking, the calling function stores the
+return address to BEAM register and the called function pushes that
+register to the stack.)
+
+Calls to other modules were done similarly by using the GCC extension
+making it possible to take the address of a label and later jumping to
+it.  Thus an external call was made by pushing the return address to
+the stack followed by `goto` to the address of a label in another C
+function.
+
+Isn't that undefined behavior?
+
+Yes, it is undefined behavior even in GCC. It happened to work with
+GCC on Sparc, but not on GCC for X86. A further complication was the
+embedded systems having ANSI-C compilers without any GCC extensions.
+
+Because of that, we had to maintain three distinct flavors of BEAM/C
+to handle different C compilers and platforms. I don't remember any
+benchmarks from that time, but it is unlikely that BEAM/C was faster
+than interpreted BEAM on any other platform than Solaris on Sparc.
+
+In the end, we removed BEAM/C and optimized the interpreted BEAM so that
+it could beat BEAM/C in speed.
+
+### HiPE
+
+[HiPE][hipe] (The High-Performance Erlang Project) was a research
+project at Uppsala University running for many years starting around
+1996. It was "aimed at efficiently implementing concurrent programming
+systems using message-passing in general and the concurrent functional
+language Erlang in particular".
+
+[hipe]: https://www.it.uu.se/research/group/hipe/
+
+One of the many outcomes of the project was the HiPE native code
+compiler for Erlang. HiPE became a part of the OTP distribution in OTP
+R8 in 2001.  The HiPE native compiler is written in Erlang and
+translates the BEAM code to native code without the help of a C
+compiler, therefore avoiding many of the problems that BEAM/C ran into.
+
+The HiPE native compiler can often speed up sequential code by a
+factor of two or three compared to interpreted BEAM code.  We hoped
+that would speed up real-world huge application systems.
+Unfortunately, projects within Ericsson that tried HiPE found that it
+did not improve performance.
+
+Why is that?
+
+The main reason is probably that many typical huge Erlang applications
+don't contain enough sequential code that HiPE could optimize.  The
+runtime of those systems is typically dominated by some combination of
+message passing, calls to the [ETS BIFs][ets], and garbage collection, none
+of which HiPE can optimize.
+
+[ets]: http://erlang.org/doc/man/ets.html
+
+Another reason could be that many systems have many small modules.  The
+HiPE native compiler (in common with the Erlang compiler) cannot optimize
+code across module boundaries, thus being unable to do much type-based
+optimizations.
+
+Also for most systems, it was not possible to compile all code to
+native code.  There is a small cost of switching from native code to
```

Comment:
maybe replace small cost with "small extra cost" or "small overhead" switching from native code to interpreted BEAM and vise versa.